packages/kit/src/utils/routing.js

```diff
@@ -21,17 +21,26 @@ export function parse_route_id(id) {
 			: new RegExp(
 					`^${get_route_segments(id)
 						.map((segment, i, segments) => {
-							const decoded_segment = decodeURIComponent(segment);
+							if (segment.includes('%')) {
+								throw new Error(`Routes ${id} must be provided in decoded form.`);
+							}
+							if (/[#\/\?]/.test(segment)) {
+								// #, ?, and / will not be decoded by decodeURI so cannot be matched
+								// '/' is illegal on all file systems so we don't bother warning about it
+								throw new Error(
+									`Route ${id} should not contain '#' or '?'. Please use a parameter.`
+								);
+							}
```

Comment:
this should happen sooner, in `create_manifest_data`. if it's here then this code will be sent to the client, needlessly